source/sigv4.c

```diff
@@ -793,6 +854,230 @@ static SigV4Status_t getCredentialScope( SigV4Parameters_t * pSigV4Params,
+
+        for( noOfHeaders = 0; noOfHeaders < headerCount; noOfHeaders++ )
+        {
+            if( canonicalRequest->pHeadersLoc[ noOfHeaders ].key.pData != NULL )
+            {
+                sigV4Status = writeCanonicalHeaderToString( noOfHeaders, canonicalRequest );
+
+                if( sigV4Status != SigV4Success )
+                {
+                    break;
+                }
+            }
+        }
+
+        return sigV4Status;
+    }
+
+    static SigV4Status_t generateCanonicalHeaders( const char * pHeaders,
+                                                   size_t headersLen,
+                                                   uint32_t flags,
+                                                   CanonicalContext_t * canonicalRequest )
+    {
+        const char * start;
+        const char * end;
+        size_t fieldFlag = 1, noOfHeaders = 0, i = 0;
+        SigV4Status_t sigV4Status = SigV4Success;
+
+        assert( pHeaders != NULL );
+        assert( canonicalRequest != NULL );
+        assert( canonicalRequest->pBufCur != NULL );
+
+        start = end = pHeaders;
+
+        for( i = 0; i < headersLen; i++ )
+        {
+            if( noOfHeaders == SIGV4_MAX_HTTP_HEADER_COUNT )
+            {
+                sigV4Status = SigV4MaxHeaderPairCountExceeded;
+            }
+
+            /* Extracting each header key and value from the headers string. */
+            if( fieldFlag == 1 )
+            {
+                if( pHeaders[ i ] == ':' )
+                {
+                    canonicalRequest->pHeadersLoc[ noOfHeaders ].key.pData = start;
+                    canonicalRequest->pHeadersLoc[ noOfHeaders ].key.dataLen = ( end - start );
+                    start = end + 1U;
+                    fieldFlag = 0;
+                }
+            }
+            else
+            {
+                if( ( ( flags & SIGV4_HTTP_PATH_IS_CANONICAL_FLAG ) && ( pHeaders[ i ] == '\n' ) ) || ( !( flags & SIGV4_HTTP_PATH_IS_CANONICAL_FLAG ) && ( pHeaders[ i ] == '\r' ) && ( ( i + 1 ) < headersLen ) && ( pHeaders[ i + 1 ] == '\n' ) ) )
+                {
+                    canonicalRequest->pHeadersLoc[ noOfHeaders ].value.pData = start;
+                    canonicalRequest->pHeadersLoc[ noOfHeaders ].value.dataLen = ( end - start );
+                    start = end + 1U;
+                    fieldFlag = 1;
+                    noOfHeaders + 1U;
+                }
+            }
+
+            end++;
+        }
+
+        /* Sorting headers based on keys. */
+        if( ( sigV4Status == SigV4Success ) && !( flags & SIGV4_HTTP_PATH_IS_CANONICAL_FLAG ) )
+        {
+            qsort( canonicalRequest->pHeadersLoc, noOfHeaders, sizeof( SigV4KeyValuePair_t ), cmpField );
+        }
+
+        if( ( sigV4Status == SigV4Success ) && !( flags & SIGV4_HTTP_PATH_IS_CANONICAL_FLAG ) )
+        {
+            sigV4Status = appendCanonicalizedHeaders( noOfHeaders, canonicalRequest );
+        }
+
```

Comment:
There must be a newline between the canonicalized headers and signed headers.

Reply:
This was done for testing things locally. While integration we need to decide whether we need to call it separately or as a part of this function based on that we can add the new line character.